src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp

```diff
@@ -3039,6 +3039,30 @@ void KviIrcServerParser::parseNumericSaslFail(KviIrcMessage * msg)
 		pOut->output(KVI_OUT_SERVERINFO, __tr2qs("SASL authentication error: %Q"), &szParam);
 	}
 
+	// Handle fallback if possible for SASL auth failure or dump if user mandates SASL
+	// and no fallback is available
+	if(msg->numeric() == 904)
+	{
+		if(msg->connection()->stateData()->sentSaslMethod() == QStringLiteral("EXTERNAL"))
+		{
+			if(!msg->connection()->target()->server()->saslNick().isEmpty() && !msg->connection()->target()->server()->saslPass().isEmpty())
+			{
+				KviWindow * pOut = static_cast<KviWindow *>(msg->console());
+				pOut->output(KVI_OUT_SERVERINFO, __tr2qs("Attempting fallback due to SASL failure."));
+				msg->connection()->sendFmtData("AUTHENTICATE PLAIN");
+				msg->connection()->stateData()->setSentSaslMethod(QStringLiteral("PLAIN"));
+				return;
+			}
+		}
+
+		if(KVI_OPTION_BOOL(KviOption_boolDropConnectionOnSaslFailure))
+		{
+			KviWindow * pOut = static_cast<KviWindow *>(msg->console());
+			pOut->output(KVI_OUT_SERVERINFO, __tr2qs("SASL auth failed. Dropping the connection."));
+			msg->connection()->sendFmtData("QUIT");
```

Comment:
Server isn't guaranteed to disconect on QUIT though, even if it probably will... Can we send QUIT and then disconnect?

Reply:
The idea is to not force a disconnect so that it will automatically attempt to connect again, in case services are having issues, but I suppose in most situations where services would be having issues, you wouldn't get a 904 response.